### PR TITLE
Feature : Allow to broadcast RN side

### PR DESCRIPTION
**Pull request by seba9999**
Hi @kirillzyusko,

I've worked a bit on that but I don't have any clue on how to compile this into a react native library ! 
Could you explain me a bit ? Or compile it and let me try it in react native to see if it works ?

Thanks :)

### CODE REVIEW

File: android/src/main/java/io/wifi/p2p/WiFiP2PManagerModule.java [hunk 1]
```diff
@@ -304,16 +316,18 @@ protected void onReceiveResult(int resultCode, Bundle resultData) {
         getCurrentActivity().startService(serviceIntent);
     }
 
+
```
Comment (kirillzyusko) [inline android/src/main/java/io/wifi/p2p/WiFiP2PManagerModule.java:319]: Why you've added these empty lines?

**Comment by kirillzyusko** [comment]
Hi @seba9999 

Yes, sure. I've used https://github.com/kirillzyusko/react-native-wifi-p2p-example example app to test my changes. There you can specify a package from github:

```json
"react-native-wifi-p2p": "https://github.com/seba9999/react-native-wifi-p2p#feature-broadcast"
```

Or you can use [symlink](https://apple.stackexchange.com/a/115647) to link a local package (I prefer this option, since I can easily modify file locally and I don' need to reinstall deps every time).

Let me know if you have any other questions!

BTW, you changes look good for me! I'll be happy to approve/merge/release a new version/close the issue if you confirm these changes actually works for you and allow you to achieve what you want 😎

**Comment by kirillzyusko** [comment]
Also @seba9999 I'd like to ask you to document new method `sendMessageTo` in README.

And maybe new method for sending file also can be added? `sendFileTo`?

File: android/src/main/java/io/wifi/p2p/WiFiP2PManagerModule.java (hunk 1 above)
Comment (seba9999) [reply android/src/main/java/io/wifi/p2p/WiFiP2PManagerModule.java]: deleted :)

**Comment by seba9999** [comment]
I've updated the code and the docs :) 

I've also added a sendFileTo ;) 

Sorry for messing your README.md file but doesn't it look prettier now ? :P 

For the testing part I'm a bit blocked... I've never used react native alone .. I'm using [expo](https://expo.dev/) so I don't really get it when it comes to symlink ... 
Could you elaborate ? 
Do I have to symlink my index.js ( of this repo ) into my  `/nodes_modules` of the repo of my example ? 
The difficult part is that in order to test specific library with expo I must build a [developement build](https://docs.expo.dev/development/create-development-builds/) first so I don't think this gonna work ...

**Comment by kirillzyusko** [comment]
@seba9999 you need to symlik this repo to `node_modules`. After that you should see a folder `react-native-wifi-p2p` inside of `node_modules` - in fact it will be just a reference to an actual folder. And build the app. That's how it works for pure RN projects. I worked a little bit with expo dev client, and I think you don't need to change anything (maybe only be sure, that you have `react-native-wifi-p2p` package in `dependencies` - maybe linker somehow consumes this info).

Regarding prettier part - yes, it looks much better, but I'd like to keep changes separately. Would you mind to submit changes without prettier and when this PR will be merged create another PR with prettier integration (also maybe add github actions to assure all code is properly formatted). What do you think about it? In this cases changes will have better git-history and will not be mixed inside of one PR (which menas it will be easier for me to review them).

File: index.js [hunk 2]
```diff
@@ -53,92 +54,106 @@ const subscribeOnConnectionInfoUpdates = (callback) => subscribeOnEvent(CONNECTI
...
     });
-});
+  });
 
 const getConnectionInfo = () => WiFiP2PManager.getConnectionInfo();
 
 const getGroupInfo = () => WiFiP2PManager.getGroupInfo();
 
 export {
-    // public methods
```
Comment (kirillzyusko) [inline index.js:113]: Could you also revert prettier changes in this file?

File: README.MD [hunk 3]
```diff
@@ -332,6 +335,19 @@ You should call this method if you want to send from client side to server (grou
...
+
+You should call this method if you want to send message to a specific address (peer address in the same group). Before sending you should execute next steps:
+
+1. Create group
+2. Connect to group from client side
+3. After establishing connection you should call `getConnectionInfo()` on client side (also recommend do it on group-server side)
+4. Call `receiveMessage()` on at least one side (`receiveMessage` return `Promise<string>` - message from client)
+5. Call `sendMessage(message)` on on other side, after resolving `getConnectionInfo()`
+
+_Note_: you cannot send character encoding for string and by default this library uses `UTF-8`.
```
Comment (kirillzyusko) [inline README.MD:349]: Maybe better to write?

`Same function as above but you can specify the address (a peer address in the same group as this device is).`

**Comment by kirillzyusko** [comment]
Also @seba9999 don't you think that we should to port these changes?

https://github.com/coletivoEITA/react-native-wifi-p2p-reborn/blob/b657850b8034a298dc71334ffc20a3f346d467c1/android/src/main/java/io/wifi/p2p/MessageServerAsyncTask.java#L61-L63

File: index.js (hunk 2 above)
Comment (seba9999) [reply index.js]: Didn't see ^^' I've changed :)

File: index.js [hunk 4]
```diff
@@ -99,6 +101,8 @@ const receiveFile = (folder, fileName, forceToScanGallery = false) => new Promis
 
 const sendMessage = (message) => WiFiP2PManager.sendMessage(message);
 
+const sendMessageTo = (message, address) => WiFiP2PManager.sendMessageTo(message, address);
```
[the remark this answers is not in the document]
Comment (kirillzyusko) [reply index.js:104]: @seba9999 and please, add methods definition in `index.d.ts` 🙏

[the remark this answers is not in the document]
Comment (seba9999) [reply index.js:104]: There isn't an index.d.ts in this repo :/

**Comment by kirillzyusko** [comment]
Looks good 👍 
I'm going to merge.

One more question @seba9999 - are you planning to change README? Add information about Android 13, setup prettier etc.?

**Comment by seba9999** [comment]
I'll may do it if this effectively works ! 
Are you creating the 3.3.1 ( or 3.4.0 )  release ?

**Comment by kirillzyusko** [comment]
@seba9999 I will create 3.4.0 release, since this release will contain API enhancements 👀

**Comment by seba9999** [comment]
Hum I've finally managed to test it out but I'm facing a problem ...
The new "address" parameter I've created is waiting for an IP address and in the list of peer we only have their MAC Addresses ! ...

I think we need some refactoring in order to store manually the IPs of all the devices in the group ... Allowing the "broadcast system" in the library itself ...

I didn't find something newer than this [10 years old thread](https://stackoverflow.com/questions/10053385/how-to-get-each-devices-ip-address-in-wi-fi-direct-scenario) ... ><

**Comment by kirillzyusko** [comment]
@seba9999 contributions are always welcome :)

And feel free to create a new issues, since someone may need this information, but it will be hard for them to find a pieces of informations in comments under PR 😅

**Pull request merged**